src/chk/chk_internal.h

```diff
@@ -40,8 +46,119 @@ enum chk_rpc_opc {
+#define CHK_BK_LEADER		"leader"
+
+/* akey for engine bookmark under CHK_DB_TABLE */
+#define CHK_BK_ENGINE		"engine"
+
+/* akey for check property under CHK_DB_TABLE */
+#define CHK_PROPERTY		"property"
+
+#define CHK_BK_MAGIC_LEADER	0xe6f703da
+#define CHK_BK_MAGIC_ENGINE	0xe6f703db
+#define CHK_BK_MAGIC_POOL	0xe6f703dc
+
+/*
+ * XXX: Please be careful when change CHK__CHECK_INCONSIST_CLASS__CIC_UNKNOWN
+ *	to avoid hole is the struct chk_property.
+ */
+#define CHK_POLICY_MAX		(CHK__CHECK_INCONSIST_CLASS__CIC_UNKNOWN + 1)
+#define CHK_POOLS_MAX		(1 << 6)
+#define CHK_RANKS_BITMAP_MAX	(1 << 24)
+
+/*
+ * Each check instance has a unique leader engine that uses key "chk/leader" under its local
+ * sys_db to trace the check instance.
+ *
+ * For each engine, include the leader engine, there is a system level key "chk/engine" under
+ * the engine's local sys_db to trace the check instance on the engine. When server (re)start
+ * the check module uses it to determain whether needs to rejoin the check instance.
+ *
+ * For each pool, there is a key "chk/$pool_uuid" under the engine's local sys_db to trace
+ * check process for the pool on related engine.
+ */
+struct chk_bookmark {
+	uint32_t			cb_magic;
+	uint32_t			cb_version;
+	uint64_t			cb_gen;
+	Chk__CheckScanPhase		cb_phase;
+	union {
+		Chk__CheckInstStatus	cb_ins_status;
+		Chk__CheckPoolStatus	cb_pool_status;
+	};
+	/*
+	 * For leader bookmark, it is the inconsistency statistics during the phases range
+	 * [CSP_PREPARE, CSP_POOL_LIST] for the whole system. The inconsistency and related
+	 * reparation during these phases may be in MS, not related with any engine.
+	 *
+	 * For pool bookmark, it is the inconsistency statistics during the phases range
+	 * [CSP_POOL_MBS, CSP_CONT_CLEANUP] for the pool. The inconsistency and related
+	 * reparation during these phases is applied to the pool service leader.
+	 */
+	struct chk_statistics		cb_statistics;
+	struct chk_time			cb_time;
+};
+
+/* On each engine (including the leader), there is a key "chk/property" under its local sys_db. */
+struct chk_property {
+	d_rank_t			cp_leader;
+	Chk__CheckFlag			cp_flags;
+	Chk__CheckInconsistAction	cp_policies[CHK_POLICY_MAX];
+	/*
+	 * How many pools will be handled by the check instance. -1 means to handle all pools.
+	 * If the specified pools count exceeds CHK_POOLS_MAX, then all pools will be handled.
+	 */
+	int32_t				cp_pool_nr;
+	uuid_t				cp_pools[CHK_POOLS_MAX];
+	/*
+	 * XXX: Preserve for supporting to continue the check until the specified phase in the
+	 *	future. -1 means to check all phases.
+	 */
+	int32_t				cp_phase;
+	/* How many ranks (ever or should) take part in the check instance. */
+	uint32_t			cp_rank_nr;
+	/* The size of cp_ranks_bitmap in byte. */
+	uint32_t			cp_bitmap_sz;
+	/*
+	 * Bitmap for the ranks that participate in the check instance. It is at most
+	 * CHK_RANKS_BITMAP_MAX (16M) bytes, because leader bookmark only exists on the
```

Comment:
Why do we need such a large bitmap? 16M bytes can contain 128 million ranks? I think a 16k bitmap should be large enough.

Reply:
I will adjust the patch a bit to dynamically calculate and allocate bitmap to avoid large bitmap.